### PR TITLE
feat: prototype pelikan_rds server

### PR DESCRIPTION
**Pull request by hderms**
*NOT* ready for production:

This implements `get` and `set` commands for the RESP protocol and permits starting an 'Rds' server (example: `cargo run --bin pelikan_rds`)

There is an example configuration file at `config/rds.toml`

### CODE REVIEW

File: src/protocol/resp/src/lib.rs [hunk 1]
```diff
@@ -1,16 +1,21 @@
-// Copyright 2022 Twitter, Inc.
+// Copyright 2023 Pelikan Foundation LLC.
```
Comment (brayniac) [inline src/protocol/resp/src/lib.rs:1]: I'm not sure we can change the copyright on this. I believe we can add an additional 2023 copyright, but I don't think we can remove the current Twitter one. CC @thinkingfish

File: src/server/rds/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+        // initialize metrics
+        common::metrics::init();
+
+        // initialize storage
+        let storage = Storage::new(&config)?;
+
+        // initialize parser
+        let parser = Parser::new();
+        // .max_value_size(config.seg().segment_size() as usize)
```
Comment (brayniac) [inline src/server/rds/src/lib.rs:36]: remove commented out lines

File: src/server/rds/tests/integration.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2021 Twitter, Inc.
```
Comment (brayniac) [inline src/server/rds/tests/integration.rs:1]: I think this can be 2023 Pelikan Foundation

File: src/protocol/resp/src/lib.rs (hunk 1 above)
Comment (thinkingfish) [reply src/protocol/resp/src/lib.rs:1]: +1, we should add a new line

File: src/protocol/resp/src/request/mod.rs [hunk 4]
```diff
@@ -59,10 +62,15 @@ pub use hset::*;
 pub use hvals::*;
 pub use sadd::*;
 pub use set::*;
+// response codes for klog
+const MISS: u8 = 0;
```
Comment (brayniac) [inline src/protocol/resp/src/request/mod.rs:66]: I feel like I'd prefer these to be enum variants with associated values.

File: src/protocol/resp/src/request/mod.rs [hunk 5]
```diff
@@ -59,10 +62,15 @@ pub use hset::*;
...
+const HIT: u8 = 4;
+const STORED: u8 = 5;
+const NOT_STORED: u8 = 9;
```
Comment (brayniac) [inline src/protocol/resp/src/request/mod.rs:69]: I don't think this is the right code to use for what we're using it for? `NOT_STORED` is the result of conditions for add/replace not being met. See: https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L277

File: src/protocol/resp/src/request/set.rs [hunk 6]
```diff
@@ -235,6 +247,26 @@ impl Compose for Set {
     }
 }
 
+impl Klog for Set {
+    type Response = Response;
+
+    fn klog(&self, response: &Self::Response) {
+        let code = match response {
+            Message::SimpleString(_) => STORED,
+            _ => NOT_STORED,
```
Comment (brayniac) [inline src/protocol/resp/src/request/set.rs:256]: It seems like we need to use an error result code here.

File: src/server/rds/benches/benchmark.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+//! This is a very basic benchmark which tests only get requests with a few
+//! different key and value sizes. It's only using one connection and a very
+//! primitive blocking client, so these results do not reflect the true
+//! performance of the server when under load. It can be used to get a rough
+//! idea of how changes may impact performance.
+//!
+//! For formal performance testing, it is recommended to use
+//! [rpc-perf](https://github.com/twitter/rpc-perf) or another cache
+//! benchmarking tool which supports the Memcache ASCII protocol.
```
Comment (brayniac) [inline src/server/rds/benches/benchmark.rs:13]: `s/Memcache ASCII/RESP/`

File: src/server/rds/benches/benchmark.rs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+use pelikan_rds::Rds;
+
+use std::io::{Read, Write};
+use std::net::TcpStream;
+use std::time::Duration;
+const RESP_NIL: &[u8; 5] = b"$-1\r\n";
+const RESP_OK: &[u8; 5] = b"+OK\r\n";
+fn get_benchmark(c: &mut Criterion) {
+    // use the default config
+    let config = SegcacheConfig::default();
```
Comment (brayniac) [inline src/server/rds/benches/benchmark.rs:26]: It probably shouldn't be called `SegcacheConfig` here... but okay to address in a separate PR.

File: src/server/rds/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+/// This structure represents a running `Rds` process.
+#[allow(dead_code)]
+pub struct Rds {
+    process: Process,
+}
+
+impl Rds {
+    /// Creates a new `Rds` process from the given `SegcacheConfig`.
+    pub fn new(config: SegcacheConfig) -> Result<Self, std::io::Error> {
```
Comment (brayniac) [inline src/server/rds/src/lib.rs:24]: Again, same thing here.

File: src/server/rds/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 Pelikan Foundation LLC.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+//! Rds is an implementation of a cache backend that implements a subset of
+//! the Redis ASCII protocol and is currently backed with segment based storage. By
```
Comment (brayniac) [inline src/server/rds/src/main.rs:6]: `s/Redis ASCII protocol/Redis commands/`?

File: src/server/rds/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+extern crate logger;
+
+use backtrace::Backtrace;
+use clap::{Arg, Command};
+use config::SegcacheConfig;
+use metriken::*;
+use pelikan_rds::Rds;
+use server::PERCENTILES;
+
+/// The entry point into the running Segcache instance. This function parses the
```
Comment (brayniac) [inline src/server/rds/src/main.rs:25]: s/Segcache/RDS/

File: src/server/segcache/benches/benchmark.rs [hunk 12]
```diff
@@ -36,7 +36,7 @@ fn get_benchmark(c: &mut Criterion) {
     let mut buffer = vec![0; 1024 * 1024];
 
     // define a benchmarking group
-    let mut group = c.benchmark_group("request");
+    let mut group = c.benchmark_group("request_seg");
```
Comment (brayniac) [inline src/server/segcache/benches/benchmark.rs:39]: Why the change here?

**Review by brayniac — CHANGES_REQUESTED** [review]

File: src/server/segcache/benches/benchmark.rs (hunk 12 above)
Comment (hderms) [reply src/server/segcache/benches/benchmark.rs:39]: I may have misinterpreted a need to have the bench names be unique to also extend to benchmark groups, which probably isnt the case. I can change it back

File: src/protocol/resp/src/request/mod.rs (hunk 5 above)
Comment (hderms) [reply src/protocol/resp/src/request/mod.rs:69]: @brayniac do you think preserving the memcached meanings in Klog would be helpful? I assumed there was a chance it would be better than defining an entirely new set of values just for RESP, mostly to make it simpler for endusers, but I hadn't really asked that question.

Is Klog something that someone would actually run in prod, anyways? My preference would be for Klog to render the code as a full string, but I could see how the performance of that would be impossible to justify. Right now we have like, something that's sort of a compact textual representation (i.e. not binpacked or anything but with minimal characters being logged). Curious if this would be considered a performance sensitive enough area of the codebase to preclude logging anything but a single character

File: src/server/rds/benches/benchmark.rs (hunk 8 above)
Comment (hderms) [reply src/server/rds/benches/benchmark.rs:26]: yeah I can possibly fix this in this PR. Let me check what it would require

**Review by brayniac — APPROVED** [review]
LGTM
